core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java

```diff
@@ -1323,6 +1326,42 @@ public boolean usesLegacyTableLayouts(Session session, TableHandle table)
                 });
     }
 
+    @Override
+    public Optional<JoinApplicationResult<TableHandle>> applyJoin(
+            Session session,
+            JoinType joinType,
+            TableHandle left,
+            TableHandle right,
+            List<JoinCondition> joinConditions,
+            Map<String, ColumnHandle> leftAssignments,
+            Map<String, ColumnHandle> rightAssignments)
+    {
+        if (!right.getCatalogName().equals(left.getCatalogName())) {
+            return Optional.empty();
+        }
+        CatalogName catalogName = left.getCatalogName();
+
+        ConnectorTransactionHandle transaction = left.getTransaction();
+        ConnectorMetadata metadata = getMetadata(session, catalogName);
+        ConnectorSession connectorSession = session.toConnectorSession(catalogName);
+
+        Optional<JoinApplicationResult<ConnectorTableHandle>> connectorResult = metadata.applyJoin(connectorSession, joinType, left.getConnectorHandle(), right.getConnectorHandle(), joinConditions, leftAssignments, rightAssignments);
+
+        connectorResult.ifPresent(result -> {
```

Comment:
nit: i think i would prefer the validation to be part of .map call, as in applyAgg above